macro-mathjax-ui/src/main/resources/Macros/MathJaxMacro.xml

```diff
@@ -136,40 +136,74 @@ y = \frac{c}{cb-ad}
-  // to empty pages when exporting to PDF even when no message is displayed because they are not hidden (height: 1px).
-  // The best practice is to inject such DIVs at the end of the BODY element, which this clean-up function does.
-  const cleanUp = function() {
-    // Move the hidden MathJax containers at the end of the BODY element.
-    const containers = document.querySelectorAll('body &gt; div &gt; #MathJax_Hidden, body &gt; div &gt; #MathJax_SVG_Hidden');
-    containers.forEach(function(container) {
-      document.body.appendChild(container.parentNode);
-    });
+      <code>define('configMathjax', function() {
+  // See http://docs.mathjax.org/en/latest/configuration.html
+  window.MathJax = {
+    chtml: {
+      displayAlign: "left",
+      displayIndent: "2em"
+    },
+    startup: {
+      elements: document.getElementsByClassName("xwiki-mathjax")
+    },
+    tex: {
+      tags: "ams",
+      autoload: {
+        color: [],
+        colorv2: ['color']
+      },
+      packages: {'[+]': ['noerrors']}
+    },
+    options: {
+      ignoreHtmlClass: 'tex2jax_ignore',
+      processHtmlClass: 'tex2jax_process'
+    },
+    loader: {
+      load: ['[tex]/noerrors']
+    }
   };
+});
+
+// Configure MathJax before loading it.
+require.config({
+  paths: {
+    'mathjax': "$services.webjars.url('org.webjars.npm:mathjax', 'es5/tex-chtml.js')"
+  },
+  shim: {
+    'mathjax': ['configMathjax']
+  }
+});
 
+require(["jquery", 'mathjax'], function($, math) {
   // Delay the page ready until all formulas are generated, if supported.
   if (require.defined('xwiki-page-ready')) {
     require(['xwiki-page-ready'], function(pageReady) {
       pageReady.delayPageReady(new Promise(function(resolve, reject) {
-        // See http://docs.mathjax.org/en/v2.7-latest/advanced/queues.html#the-mathjax-processing-queue
-        MathJax.Hub.Queue(function() {
-          cleanUp();
+        // See https://docs.mathjax.org/en/latest/web/configuration.html#performing-actions-after-typesetting
+        MathJax.startup.promise.then(() =&gt; {
           resolve();
         });
       }), 'MathJax');
     });
   }
+
+  // Chain typesetting calls, since multiple simultaneously asynchronous typesetting calls can produce errors.
+  // See https://docs.mathjax.org/en/latest/web/typeset.html#handling-asynchronous-typesetting.
+  let promise = Promise.resolve();
+  var typeset = function(elems) {
+    promise = promise.then(() =&gt; {
+      // Reset the tex labels, since re-typesetting might create duplicate labels.
+      MathJax.texReset();
+      MathJax.typesetPromise(elems);
+    }).catch((err) =&gt; console.log('Typeset failed: ' + err.message));
+  };
+
+  $(document).on('xwiki:dom:updated', function(e, data) {
+    let mathElems = data.elements.flatMap(x =&gt; {
+      return x.hasClassName('xwiki-mathjax') ? [x] : [...x.querySelectorAll('.xwiki-mathjax')];
+    });
+    typeset(mathElems);
+  });
```

Comment:
What do you think of initializing the promise with the promise of `xwiki-page-ready` is it is defined? So that the rest of the typesetting is done after the page is ready? Not fully sure it's required (and it's not tested).

```javascript
require(["jquery", 'mathjax'], function ($) {
  // Delay the page ready until all formulas are generated, if supported.
  var promise = Promise.resolve();
  if (require.defined('xwiki-page-ready')) {
    // Override the promise when xwiki-page-ready is defined so that the rest of the typesettings is done after the 
    // page is ready.
    promise = new Promise(function (resolve) {
      // See https://docs.mathjax.org/en/latest/web/configuration.html#performing-actions-after-typesetting
      MathJax.startup.promise.then(() =&gt; {
        resolve();
      });
    })
    require(['xwiki-page-ready'], function (pageReady) {
      pageReady.delayPageReady(pageReadyPromise, 'MathJax');
    });
  }

  // Chain typesetting calls, since multiple simultaneously asynchronous typesetting calls can produce errors.
  // See https://docs.mathjax.org/en/latest/web/typeset.html#handling-asynchronous-typesetting.

  var typeset = function (elems) {
    promise = promise.then(() =&gt; {
      // Reset the tex labels, since re-typesetting might create duplicate labels.
      MathJax.texReset();
      MathJax.typesetPromise(elems);
    }).catch((err) =&gt; console.log('Typeset failed: ' + err.message));
  };

  $(document).on('xwiki:dom:updated', function (e, data) {
    let mathElems = data.elements.flatMap(x =&gt; {
      return x.hasClassName('xwiki-mathjax') ? [x] : [...x.querySelectorAll('.xwiki-mathjax')];
    });
    typeset(mathElems);
  });
})
```

Reply:
I think that it could be useful. This way we can also be sure that the subsequent typesetting calls are done after the initial one